app/models/test_run.rb

```diff
@@ -28,50 +28,56 @@ def failure(problems)
 
   def update_results!(results)
     if results['status'] == 'failed'
-      failure(results['error'])
-    else
-      self.update!(status: :complete, problems: results['error'])
-      new_overall_comments = []
-      results['test_groups'].each do |result|
-        test_group_result = create_test_group_result(result)
-        marks_total, marks_earned = 0, 0
-        result['tests'].each_with_index do |test, i|
-          ApplicationRecord.transaction do
-            test_group_result.test_results.create!(
-              name: test['name'],
-              status: test['status'],
-              marks_earned: test['marks_earned'],
-              output: test['output'].gsub("\x00", '\\u0000'),
-              marks_total: test['marks_total'],
-              time: test['time'],
-              position: i + 1
-            )
-            marks_earned += test['marks_earned']
-            marks_total += test['marks_total']
-          rescue StandardError => e
-            extra_info = test_group_result.extra_info
-            test_name = test['name'].nil? ? '' : "#{test['name']} - "
-            if extra_info.nil?
-              test_group_result.update(extra_info: test_name + e.message)
-            else
-              test_group_result.update(extra_info: extra_info + "\n" + test_name + e.message)
-            end
+      return failure(results['error'])
+    end
+    self.update!(status: :complete, problems: results['error'])
```

Comment:
This PR should not reformat the existing code. The only change that's required is 72-75.

The other change in logic (updating `marks_earned`) should not happen, similar to my comment on the markus-autotesting PR. The `add_extra_marks` function does not need to return anything, just create the `ExtraMark` objects.

Reply:
Done.